### ISSUE TITLE
Error when launching and later crash

### ISSUE THREAD

**Issue by lsaranto** [axcore/tartube#217]
Thanks for taking the time to submit a bug report! Please include the following information:

### What operating system are you using?
Linux Mint on a VM.

### What version of Tartube are you using?
2.3.0

### What happens when you run Tartube from a terminal window?
I get this error:
(tartube:812937): Gdk-CRITICAL **: 18:57:23.059: gdk_window_thaw_toplevel_updates: assertion 'window->update_and_descendants_freeze_count > 0' failed

Tartube still starts and seems to be running ok.

Later (hours, often days) Tartube might crash with error:
ython3: ../../../../src/cairo-surface.c:930: cairo_surface_reference: Assertion `CAIRO_REFERENCE_COUNT_HAS_REFERENCE (&surface->ref_count)' failed.
Aborted (core dumped)

I always lose the download db, because the file is locked when starting Tartube. It's locked even after VM reboot.

**Comment by axcore** [comment]
Tartube (and presumably the underlying Gtk graphics libraries) are not well-tested on virtual machines. This is a general problem; graphics might work, or they might not.

You see two errors. The first is caused by the Tartube icon added to the taskbar. It is an ancient GTK bug that has not been fixed on every system (it occurs in Linux Mint Cinnamon, but not Mate). If it worries you, you can disable the icon in **Edit > System preferences... > Windows > System tray**.

The second I haven't seen before. You could narrow down the possible causes by turning on debug messages, then running Tartube from within a terminal window, so we can see where the programme crashes (**Edit > System preferences... General > Debugging**). Placing an empty file called `debug.txt`, in the same directory as the executable `tartube` file, has the same effect.

As a workaround, you can click the **Check all** button first. When that has finished (the database is saved automatcailly), you can download files. youtube-dl automatically resumes partially-downloaded videos. 

When the downloads have finished, you can click **Operations > Refresh database**, which will find all the videos that have been downloaded, and modify the database accordingly.

**Comment by lsaranto** [comment]
I broke my install.

I believe I had several Tartube installs because of the many install tries in the beginning. So I wanted to remove everything and start again.

I removed everything tartube that came up with find. I think it was wrong because pip3 still tought tartube was installed. I tried to fix it by running 'sudo pip3 install --force-reinstall tartube youtube-dl'. Running tartube only showed an error.

I then uninstalled tartube with sudo pip3 uninstall tartube'.
I downloaded the deb file from releases and installed it with 'sudo apt deb python3-tartube_2.3.8.deb'.

Again, just an error when launching tartube:

```
:~$ tartube
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/tartube/mainapp.py", line 2408, in do_activate
    self.start()
  File "/usr/lib/python3/dist-packages/tartube/mainapp.py", line 2573, in start
    new_config_flag = self.load_config()
  File "/usr/lib/python3/dist-packages/tartube/mainapp.py", line 3297, in load_config
    self.thumb_size_custom = json_dict['thumb_size_custom']
KeyError: 'thumb_size_custom'
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 153, in apport_excepthook
    with os.fdopen(os.open(pr_filename,
FileNotFoundError: [Errno 2] No such file or directory: '/var/crash/_usr_bin_tartube.1000.crash'

Original exception was:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/tartube/mainapp.py", line 2408, in do_activate
    self.start()
  File "/usr/lib/python3/dist-packages/tartube/mainapp.py", line 2573, in start
    new_config_flag = self.load_config()
  File "/usr/lib/python3/dist-packages/tartube/mainapp.py", line 3297, in load_config
    self.thumb_size_custom = json_dict['thumb_size_custom']
KeyError: 'thumb_size_custom'

```

**Comment by axcore** [comment]
At this point, you would be better just running Tartube from source. 

Download the source code, then open a terminal window in the directory where the `setup.py` file is, and type

`python3 tartube/tartube`

**Comment by lsaranto** [comment]
Sorry, but no.

````
~/tartube-2.3.008$ python3 tartube/tartube
2020-12-12 20:58:33 ap 143 __init__
2020-12-12 20:58:33 app 1779 do_startup
2020-12-12 20:58:33 app 2398 do_activate
2020-12-12 20:58:33 app 2489 start
2020-12-12 20:58:33 app 6756 setup_paths
2020-12-12 20:58:33 app 6872 setup_paths
2020-12-12 20:58:33 app 6983 find_sound_effects
2020-12-12 20:58:33 mwn 103 __init__
2020-12-12 20:58:33 mwn 732 setup_pixbufs
2020-12-12 20:58:33 mwn 8839 video_catalogue_grid_reset_sizes
2020-12-12 20:58:33 app 15195 create_download_options
2020-12-12 20:58:33 app 15195 create_download_options
2020-12-12 20:58:33 app 15490 create_ffmpeg_options
2020-12-12 20:58:33 app 3127 load_config
2020-12-12 20:58:33 app 4456 get_config_path
2020-12-12 20:58:33 app 6957 convert_version
2020-12-12 20:58:33 app 6957 convert_version
Traceback (most recent call last):
  File "/home/lasse/tartube-2.3.008/tartube/mainapp.py", line 2408, in do_activate
    self.start()
  File "/home/lasse/tartube-2.3.008/tartube/mainapp.py", line 2573, in start
    new_config_flag = self.load_config()
  File "/home/lasse/tartube-2.3.008/tartube/mainapp.py", line 3297, in load_config
    self.thumb_size_custom = json_dict['thumb_size_custom']
KeyError: 'thumb_size_custom'

```

**Comment by axcore** [comment]
Looks like your config file is broken. You can probably find it at `/home/YOURNAME/.config/tartube/settings.json`. 

(Probably you already know that this is a hidden folder, so press CTRL+H to show hidden folders.)

If you find it there, perhaps you can show me the contents. Then you can delete it.

That's the standard location for config files. If you create an empty file called `settings.json` in the Tartube directory (the one containing the setup.py file), Tartube will use that location, instead. So you should try that, as well.

**Comment by lsaranto** [comment]
I decided to ditch the Linux Mint and created a new VM with Debian. I got Tartube 2.3.008 to run (python3 tartube/tartube), but I can't get it to download. It's just showing "Waiting..." in the Classic mode window. I'm not sure if youtube-dlc is in the right location. I installed it with 'pip3 install youtube-dlc'. It puts it in ~/.local/bin , and I've selected that option in Tartube settings. What's the "local path" option? I tried copying youtube-dlc executable to a few different locations.

I noticed this in the debug log.
```
2020-12-13 22:48:53 app 16756 on_button_classic_download
2020-12-13 22:48:53 app 7803 download_manager_start
Traceback (most recent call last):
  File "/home/lasse/tartube-2.3.008/tartube/mainapp.py", line 16781, in on_button_classic_download
    self.download_manager_start('classic_real')
  File "/home/lasse/tartube-2.3.008/tartube/mainapp.py", line 7895, in download_manager_start
    ).format(str(disk_space), str(total_space)),
TypeError: show_msg_dialogue() missing 2 required positional arguments: 'msg_type' and 'button_type'

```

I really shouldn't be using Linux.

**Comment by axcore** [comment]
Oh dear, you're not having much luck, are you? There are a few things here. If you're in a hurry, go straight to 5.

1, First, make sure youtube-dlc and not youtube-dl is the selected downloader (Edit > System preferences... Downloads > Forks). I'm sure you have already done that

2. The 'use local path' assumes that youtube-dl(c) is installed in the user's PATH, in other words, you can run youtube-dl(c) from any location without specifying its directory (e.g. open a terminal window, and type `youtube-dlc --version`)

3. If you want to know, you can find out what each **Path to the executable** option means; check or download a video and see the yellow text at the top of the **Output** tab. Likewise, you can see what each **Command for update operations** means, by clicking **Operations > Update youtube-dl(c)**, and again checking the yellow text in the **Output** tab

4. If you installed youtube-dl(c) via pip, then the **Use PyPI path *~/local/bin/youtube-dlc)** option is the corrent one

5. Finally, you're seeing a python error because a dialogue window is broken. The window should say: `You only have {0} / {1} Mb remaining on your device`. I can fix the broken window, but I can't fix the lack of space on your (virtual) hard drive.

You can disable the hard drive check. **Edit > System preferences... > Files > Device**, and deselect **Warn user if disk space is less than** and also deselect **Halt downloads if disk space is less than**.

My guess is that youtube-dlc is installed fine, Tartube can find it just fine, but nothing is happening because of the broken window.

**Comment by lsaranto** [comment]
Ah, so simple when it's laid out like that.

I've set the Destination (Classic mode) to be a mount point to my NAS. It definitely has enough free space left. You might need to check your disk space routine. Or provide a hint for a similar situation. I removed the checks and it's now downloading.

Thank you for your help and patience. I was gonna buy you a coffee or something, but couldn't find a donations link.

P.S. At one point I did the manual install (python3 setup.py install) (possibly with sudo). With the Linux Mint it worked with an older version, but on this Debian it complains about not finding icons. Anyways, it's not a problem for me as I now have a running copy.

**Comment by axcore** [comment]
I uploaded the fix. If you feel like it, you could replace the old `/tartube/mainapp.py` file with the new one (or just replace everything from github). Then we can see much disk space Tartube thinks you have.

Install with `python3 setup.py install`, using `sudo`, is not tested. Doing that without `sudo` is tested and should work (see [here](https://github.com/axcore/tartube#539-install-from-source)). Installation from PyPI is tested both with and without root privileges (see [here](https://github.com/axcore/tartube#536-install-using-pypi-with-root-privileges)).

No need for a coffee, I'm happy if people take time out of their busy days to report a fault ;)

**Comment by lsaranto** [comment]
One last thing (I hope).

I'm getting additional files (.json & .jpg) when downloading. I don't really need them. They didn't get created before. Downloading works if I put nothing as the prefered format in the (General) download options. I tried putting Any video, plain Mp4 and Mp4 720p & 1080p as formats as I had before, but then it doesn't download at all (shows Error). What's the meaning of the number between the '<' and '>'?

**Comment by lsaranto** [comment]
I downloaded the master branch as a zip file and extracted it. When I run it, in System preferences' Files / Device tab it shows a bit under 400 Mb as free space which is accurate for the actual VM's disk. However,

```
lasse@debian:~/tartube-2.3.008$ df -h|grep  /mnt
//192.168.0.1/root   20T   16T  3.6T  82% /mnt/unraid
```

It's a samba share if that is relevant info.

**Comment by axcore** [comment]
>I'm getting additional files (.json & .jpg) when downloading. I don't really need them.

There are four files for each (YouTube) video. To set which files are created during the download, click **Edit > General download options... > Files > Write/move files**, and select the **File write options**.

To set which files are kept permanently, click the **Keep files** tab, and edit those options.

>I tried putting Any video, plain Mp4 and Mp4 720p & 1080p as formats as I had before, but then it doesn't download at all (shows Error)

There is no **Any video** format, you probably meant **Any format [720p]** or something like that. Be aware that if you specify formats that don't exist, then nothing will be downloaded and you'll get an error.

If you are not sure why you are seeing an error, you can get a list of available formats for a video. Right-click the video and select **Fetch > Available formats**. Then click the **Output** tab to see the result.

>What's the meaning of the number between the '<' and '>'?

YouTube assigns a unique number to each video format; see the list generated above and it should become clear.

**Comment by axcore** [comment]
>When I run it, in System preferences' Files / Device tab it shows a bit under 400 Mb as free space which is accurate for the actual VM's disk. However,

Do you mean that Tartube is running in a virtual machine, but downloads are being stored on a separate filesystem (for example, a real external drive)?

**Comment by lsaranto** [comment]
> > When I run it, in System preferences' Files / Device tab it shows a bit under 400 Mb as free space which is accurate for the actual VM's disk. However,
> 
> Do you mean that Tartube is running in a virtual machine, but downloads are being stored on a separate filesystem (for example, a real external drive)?

Yes.

**Comment by axcore** [comment]
Ok, I will check that.

**Comment by axcore** [comment]
I have tested the disk space problem using a real Linux box and an external hard drive. I was able to confirm that the Tartube code is working correctly; **Edit > System preferences > Files > Device** sees the size of the external hard drive, not the Linux box on which Tartube is running.

I don't know why the setup you describe provides different results. Tartube is definitely asking Python for the correct filesystem, but there is no guarantee that the underlying python code knows anything about virtualisation, or that one filesystem has been mounted on another. 

I do not think this is a problem I can solve, but I will add something to the README to tell other users to turn off the warnings.
